addons/binding/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/ThirdGeneration/ThirdGenerationHandler.java

```diff
@@ -0,0 +1,767 @@
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.Mac;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.PBEKeySpec;
+import javax.crypto.spec.SecretKeySpec;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.util.StringContentProvider;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpVersion;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.library.unit.SIUnits;
+import org.eclipse.smarthome.core.library.unit.SmartHomeUnits;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.FieldNamingPolicy;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+/**
+ * The {@link ThirdGenerationHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author René Stakemeier - Initial contribution
+ */
+@NonNullByDefault
+public class ThirdGenerationHandler extends BaseThingHandler {
+
+    private ThirdGenerationInverterTypes inverterType;
+
+    // GSON handler
+    private static final Gson GSON = new GsonBuilder().setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
+            .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES).create();
+    // used for logging
+    protected final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    // base URL of the web api
+    private static final String WEB_API = "/api/v1";
+
+    /*
+     * operations used for authentication
+     */
+    private static final String AUTH_START = "/auth/start";
+    private static final String AUTH_FINISH = "/auth/finish";
+    private static final String AUTH_CREATE_SESSION = "/auth/create_session";
+
+    /*
+     * operations used for gathering process data from the device
+     */
+    private static final String PROCESSDATA = "/processdata";
+
+    /*
+     * List of available channels for the concrete device
+     */
+
+    /*
+     * After the authentication the result (the session id) is stored here and used to "sign" future requests
+     */
+    @Nullable
+    protected String sessionId = null;
```

Comment:
Regarding Nullable annotations on all fields. Write as follows:
```suggestion
    protected @Nullable String sessionId;
```

Reply:
Will be part of the next commit